### PR TITLE
  Functional tests - Add new test 'BO > Catalog > Cart rules customer group restriction

### PR DESCRIPTION
**Pull request by cfarhani06**
Questions | Answers
-- | --
Branch? | 8.0.x
Description? | BO - Cart rules - Case 12 - Customer Group Restriction
Type? | refacto
Category? | TE
BC breaks? | no
Deprecations? | no
Fixed ticket? | no
Related PRs | no

### CODE REVIEW

File: tests/UI/data/faker/cartRule.js [hunk 1]
```diff
@@ -71,6 +71,8 @@ class CartRuleData {
     /** @type {number} Amount of times a user can use the cart rule */
     this.quantityPerUser = cartRuleToCreate.quantityPerUser || 1;
 
+    this.customerGroupSelection = cartRuleToCreate.customerGroupSelection || false;
```
Comment (Progi1984) [inline tests/UI/data/faker/cartRule.js:81]: Miss a JSDoc

File: tests/UI/pages/FO/checkout/index.js [hunk 2]
```diff
@@ -245,6 +250,25 @@ class Checkout extends FOBasePage {
     await this.clickAndWaitForNavigation(page, this.paymentConfirmationButton);
   }
 
+  /**
+   * Get All tax included price
+   * @param page {Page} Browser tab
+   * @returns {Promise<number>}
+   */
+  getATIPrice(page) {
```
Comment (Progi1984) [inline tests/UI/pages/FO/checkout/index.js]:
```suggestion
  getAllTaxesIincludedPrice(page) {
```

File: tests/UI/pages/FO/checkout/index.js [hunk 3]
```diff
@@ -245,6 +250,25 @@ class Checkout extends FOBasePage {
...
+    return this.getPriceFromText(page, this.cartTotalATI, 2000);
+  }
+
+  /**
+     * Delete the discount
+     * @param page {Page} Browser tab
+     * @returns {Promise<boolean>}
+     */
```
Comment (Progi1984) [inline tests/UI/pages/FO/checkout/index.js]:
```suggestion
  /**
   * Delete the discount
   * @param page {Page} Browser tab
   * @returns {Promise<boolean>}
   */
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+// Import data
+const CartRuleFaker = require('@data/faker/cartRule');
+const ProductData = require('@data/FO/product');
+const {DefaultCustomer} = require('@data/demo/customer');
+const {Products} = require('@data/demo/products');
+
+// import test context
+const testContext = require('@utils/testContext');
+
+const baseContext = 'functional_BO_catalog_discounts_cartRules_CRUDCartRule_customerGroupRestriction';
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]:
```suggestion
const baseContext = 'functional_BO_catalog_discounts_cartRules_CRUDCartRule_groupCustomerRestriction';
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        this,
+        'testIdentifier',
+        'viewMyShop1',
+        baseContext,
+      );
+
+      page = await addCartRulePage.viewMyShop(page);
+
+      await foHomePage.changeLanguage(page, 'en');
+      const isHomePage = await foHomePage.isHomePage(page);
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]:
```suggestion
      await foHomePage.changeLanguage(page, 'en');

      const isHomePage = await foHomePage.isHomePage(page);
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+    it('should enter a valid credentials', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'enterValidCredentials', baseContext);
+
+      await loginPage.customerLogin(page, DefaultCustomer);
+
+      const isCustomerConnected = await loginPage.isCustomerConnected(page);
+      await expect(isCustomerConnected, 'Customer is not connected!').to.be.true;
+
+      const result = await foHomePage.isHomePage(page);
+      await expect(result).to.be.true;
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]:
```suggestion
      const isHomePage = await foHomePage.isHomePage(page);
      await expect(isHomePage).to.be.true;
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    it('should click on the logo of the shop', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'CheckLogoLink',
+        baseContext,
+      );
+
+      await foHomePage.clickOnHeaderLink(page, 'Logo');
+      const pageTitle = await foHomePage.getPageTitle(page);
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]:
```suggestion
      await foHomePage.clickOnHeaderLink(page, 'Logo');

      const pageTitle = await foHomePage.getPageTitle(page);
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]:
```suggestion
      await foHomePage.clickOnHeaderLink(page, 'Logo');

      const pageTitle = await foHomePage.getPageTitle(page);
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks, but good job :)

File: tests/UI/pages/FO/checkout/index.js (hunk 2 above)
Comment (cfarhani06) [reply tests/UI/pages/FO/checkout/index.js]: Do you suggest updating the function name 'getATIPrice' for all Pages/Scenarios? 
Or duplicating the existing one to create one specific, with name 'getAllTaxesIincludedPrice' to my use case?

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 5 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]: DONE

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 6 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]: DONE

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 7 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]: DONE

File: tests/UI/pages/FO/checkout/index.js (hunk 2 above)
Comment (Progi1984) [reply tests/UI/pages/FO/checkout/index.js]: Oops sorry. I didn't know it was an override. I close my comment.

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 4 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js:32]: DONE

File: tests/UI/data/faker/cartRule.js (hunk 1 above)
Comment (cfarhani06) [reply tests/UI/data/faker/cartRule.js:81]: Done

File: tests/UI/pages/FO/checkout/index.js (hunk 3 above)
Comment (cfarhani06) [reply tests/UI/pages/FO/checkout/index.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 4 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js (hunk 8 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/12_groupCustomerRestriction.js]: Done

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: tests/UI/pages/BO/catalog/discounts/add.js [hunk 9]
```diff
@@ -66,6 +66,15 @@ class AddCartRule extends BOBasePage {
...
     this.carrierRestrictionAddBTN = '#carrier_select_add';
 
+    // Customer group selection
+    this.customerGroupRestriction = '#group_restriction';
+    this.customerGroupSelection = '#group_select_2';
+    this.customerGroupCustomer = `${this.customerGroupSelection} option:nth-child(1)`;
+    this.customerGroupGuest = `${this.customerGroupSelection} option:nth-child(2)`;
+    this.customerGroupVisitor = `${this.customerGroupSelection} option:nth-child(3)`;
+    this.customerGroupRemoveButton = '#group_select_remove';
+    this.customerGroupAddBTN = '#group_select_add';
```
Comment (Progi1984) [inline tests/UI/pages/BO/catalog/discounts/add.js:76]:
```suggestion
    this.customerGroupAddButton = '#group_select_add';
```

**Review by Progi1984 — DISMISSED** [review]

**Comment by nesrineabdmouleh** [comment]
Hello @cfarhani06,
Please check the CI error (UI tests code checks / Checking doubles in steps identifiers)
Thank you!

**Comment by boubkerbribri** [comment]
Rebase done !

**Review by Progi1984 — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Pull request merged**